src/auto-evo/simulation/food_source/HeterotrophicFoodSource.cs

```diff
@@ -68,6 +68,12 @@ public override float FitnessScore(Species species)
         // Pili are much more useful if the microbe can close to melee
         pilusScore *= predatorSpeed;
 
+        // predators are less likely to use toxin against larger prey, unless they are opportunistic
+        if (preyHexSize > microbeSpeciesHexSize)
+        {
+            oxytoxyScore *= microbeSpecies.Opportunism / Constants.MAX_SPECIES_OPPORTUNISM;
+        }
```

Comment:
I'm not sure that size should play a role in oxytoxy hunting (aside from larger prey meaning larger yield, but that's more general). Is there any reason you wanted to limit the use against larger preys?

Reply:
This area of the code is a reflection of the AI; while we use a pattern (it's actually an anti-pattern, but shh) of parallel code, the Auto-evo should reflect what the AI would do, independent of if the AI is making good choices.

As to the question of the AI choosing to not engage larger prey, I made that decision based on the idea that it's still unwise to get close to things that could engulf you. The AI still doesn't know if this "prey" has enough HP and toxin resistance to ignore the attack, or enough flagellum or a lighter membrane to chase down and kill this microbe. I tried to take the safe middle ground and say that only some of the microbes will take this risk, and having the auto-evo encourage only the bold microbes to invest in more toxin.

Reply:
> it's actually an anti-pattern, but shh

Which part is the anti-pattern? The use of threads? We either use threading or make the game much slower. Which on my computer is fine but a little bit older CPUs are really going to struggle to run the compound clouds at more than 60 FPS.

Reply:
Sorry, I didn't mean parallel as in multi-threading, I meant parallel in terms of data structures. The auto-evo, in many places (which I wrote) contained code designed to loosely mimic what the AI code does, under an implicit assumption that a developer will be smart enough to not change one without changing the other. At some point, I would like to decompose the AI code into context-independent functions that can be run by both the AI and the auto evo, so that both areas of code can ask "what would this species do in this circumstance?" by reusing the same code.

Reply:
> that a developer will be **smart enough** to not change one without changing the other

That's too optimistic. I've seen PRs that change the logic and don't bother to update a comment that's a few lines above it.

Reply:
I couldn't agree more so, hence my objection to my own design. But the refactor I would like to do will be a PR in itself, about which I'm sure we will have many conversations.